### ISSUE TITLE
grpcio-tools: ImportError: No module named 'Cython'

### ISSUE THREAD

**Issue by tejasmanohar** [grpc/grpc#7555]
```
$ pip install grpcio-tools
Collecting grpcio-tools
  Using cached grpcio_tools-0.15.1.tar.gz
    Complete output from command python setup.py egg_info:
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-build-l75jpu4j/grpcio-tools/setup.py", line 113, in <module>
        protoc_ext_module(),
      File "/tmp/pip-build-l75jpu4j/grpcio-tools/setup.py", line 105, in maybe_cythonize
        from Cython import Build
    ImportError: No module named 'Cython'

    ----------------------------------------
Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-build-l75jpu4j/grpcio-tools/
You are using pip version 8.1.1, however version 8.1.2 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
```

I think it's `cython`. Just installed python plugin from source.

**Comment by nathanielmanistaatgoogle** [comment]
Can you tell us more about your Python version, platform, and architecture?

**Comment by tejasmanohar** [comment]
```
$ python --version
Python 3.5.2
$ cat /etc/*-release
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=16.04
DISTRIB_CODENAME=xenial
DISTRIB_DESCRIPTION="Ubuntu 16.04 LTS"
NAME="Ubuntu"
VERSION="16.04 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
UBUNTU_CODENAME=xenial
$ uname -a
Linux 910bece4ab16 4.4.15-moby #1 SMP Tue Jul 19 10:26:05 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
```

I previously saw this error on OS X, x86_64 as well though, but I've moved all my development into a VM since. Are you having trouble reproducing?

**Comment by nathanielmanistaatgoogle** [comment]
What happens if you `pip install --upgrade cython` when you see the problem?

**Comment by tejasmanohar** [comment]
@nathanielmanistaatgoogle That fixes it! Makes sense. Thanks.

**Comment by nathanielmanistaatgoogle** [comment]
Reopening - @soltanmm, is it expected that developers will even get to this state?

**Comment by soltanmm** [comment]
Users building from PyPI-distributed source shouldn't see this post 1.0.

**Comment by pgrosu** [comment]
@soltanmm That is good, but we're in the middle of the `gRPC 1.0.0-pre1` testing period and he seems to be getting the `0.15.1` version through `pip`, rather than the [`1.0.0rc1` version](https://pypi.python.org/pypi/grpcio-tools).  It will be too bad to miss a user-case experience with `1.0.0rc1`.

**Comment by soltanmm** [comment]
@pgrosu The `--pre` flag must be specified with recent `pip`s when one wants to install release candidates or pre-releases of any project on PyPI. Not specifying it implies the intention of *not* wanting a pre-release.

**Comment by pgrosu** [comment]
@soltanmm Thanks for helping me with this.  I thought my mind was going when I kept getting different versions during my reinstallations.  Hopefully it will help others as well, unless I'm the only one :)

**Comment by nathanielmanistaatgoogle** [comment]
@soltanmm by "post 1.0" do you mean "fixed at 1.0" or "fixed in the first release after 1.0"? What milestone should be given to this issue?

**Comment by soltanmm** [comment]
Python 3.5 wheels (for Mac OS X targets `>10.9`) are uploaded on PyPI for 1.0 and binary installations dodge the Cython requirement OP hit. Feel free to re-open if it is believed that this is insufficient.

**Comment by JoeStaines** [comment]
Im still running into this problem when trying to have grpcio-tools as part of setup_requires in the setup.py. I noticed that grpcio-health-checking also tries to get grpcio-tools by setup_requires so in a venv i tried to `python -m pip install grpcio-health-checking` and that comes up with the 'No module named Cython' error also.

Even if i try to `pip install Cython` in the venv and try again, i get `distutils.errors.DistutilsError: Setup script exited with error: Microsoft Visual C++ 9.0 is required. Get it from http://aka.ms/vcpython27`. Though this might be separate problems

**Comment by nathanielmanistaatgoogle** [comment]
@JoeStaines: with what version of `python` and with what version of `pip` are you working?

**Comment by soltanmm-google** [comment]
@JoeStaines Short answer: `pip install grpcio-tools; pip install $grpcio_health_checking_or_whatever_that_setup_requires_grpcio_tools`.
There're some awkward things about `pip` and `setuptools` and `setup_requires`. Because `setup_requires` only supports eggs, it ignores our precompiled binaries on `PyPI`, so you end up in a situation where you either need *all* the build tools available to build everything from source (e.g. `vcpython27` on Windows is one option [that doesn't work], MSYS2 is another [which works]), or you need to bypass it all and get our wheels manually.

@nathanielmanistaatgoogle @kpayson64 I guess this event lends credence to the, "Let's not do anything fancy and just error out with instructional error messages," approach.

**Comment by sebastian-schlecht** [comment]
Hey guys,

running into the same issue as @JoeStaines - I'd like to distribute my package such that it can be installed as part of a requirements.txt file without having to worry about the order. Also I would like to run the compilation of proto files during the install to avoid compiling the into the repo.

Any solution to this as of now?

**Comment by soltanmm-google** [comment]
@sebastian-schlecht And I guess you also specifically wish to not force the user to build `grpcio-tools` and `grpcio` from source?

Unfortunately the limitation of `setup_requires` means you're up a creek on this one without telling your users to install things separately. There're efforts stemming from the SciPy/NumPy folks to make the Python native-code-build-process sane (see [PEP 518](https://www.python.org/dev/peps/pep-0518/)) such that build requirements can be met without having to bend over backwards, but until `setuptools` and `pip` and such get on board there's not much we can do on our end without supplanting a large portion of `setuptools`' functionality with DIY-hacks.

**Comment by wtanaka** [comment]
@soltanmm a python 2.7 for macos 10.9 would be useful, since python 2.7 is what's default on macos 10.9

**Comment by mjhoffman65** [comment]
Just saw this error because I was running python 3.6 (`from python:3` in Dockerfile), so it was trying to build from source. Changing to python 3.5 fixes the issue (`from python:3.5`), so it can find the wheel.

**Comment by kpayson64** [comment]
@mjhoffman65 
Looks like Python 3.6 is officially released,
I've created #9445, hopefully I can get support in for the next release.
